### PR TITLE
[WIP] alter table

### PR DESCRIPTION
**Pull request by Treggats**
Currently a WIP pull request so I have a clear overview of the changes I have made so far

### CODE REVIEW

File: src/Tempest/Database/Migrations/CreateMigrationsTable.php [hunk 1]
```diff
@@ -15,14 +16,14 @@ public function getName(): string
         return '0000-00-00_create_migrations_table';
     }
 
-    public function up(): CreateTableStatement|null
+    public function up(): QueryStatement|null
```
Comment (Treggats) [inline src/Tempest/Database/Migrations/CreateMigrationsTable.php:19]: changed the return type to indicate we're doing a query. Doesn't matter what kind

File: src/Tempest/Database/QueryStatements/AlterActionStatement.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Database\QueryStatements;
+
+use Tempest\Database\DatabaseDialect;
+use Tempest\Database\QueryStatement;
+
+final readonly class AlterActionStatement implements QueryStatement
```
Comment (Treggats) [inline src/Tempest/Database/QueryStatements/AlterActionStatement.php:10]: takes an enum with the different alter table operations

File: src/Tempest/Database/QueryStatements/AlterStatement.php [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Database\QueryStatements;
+
+enum AlterStatement: string
```
Comment (Treggats) [inline src/Tempest/Database/QueryStatements/AlterStatement.php:7]: the list of alter table operations (that I know of)

File: src/Tempest/Database/QueryStatements/BelongsToStatement.php [hunk 4]
```diff
@@ -25,17 +25,24 @@ public function compile(DatabaseDialect $dialect): string
...
-                strtolower($foreignTable),
-                strtolower($localTable),
-                $localTable,
-                $localKey,
-                $foreignTable,
-                $foreignKey,
-                'ON DELETE ' . $this->onDelete->value,
-                'ON UPDATE ' . $this->onUpdate->value,
-            ),
+            DatabaseDialect::POSTGRESQL => (new ConstraintStatement(
```
Comment (Treggats) [inline src/Tempest/Database/QueryStatements/BelongsToStatement.php:28]: extract constraint into a statement class

File: src/Tempest/Database/QueryStatements/BelongsToStatement.php [hunk 5]
```diff
@@ -25,17 +25,24 @@ public function compile(DatabaseDialect $dialect): string
...
+                sprintf(
+                    'fk_%s_%s',
+                    strtolower($foreignTable),
+                    strtolower($localTable),
+                ),
+                new RawStatement( // need to let the FK statement handle this
```
Comment (Treggats) [inline src/Tempest/Database/QueryStatements/BelongsToStatement.php:34]: todo use the ForeignKey statements

File: src/Tempest/Database/QueryStatements/DropTableStatement.php [hunk 6]
```diff
@@ -12,12 +12,20 @@
...
+        private string $tableName,
+        private array $constraints = [],
     ) {
     }
 
     public function compile(DatabaseDialect $dialect): string
     {
-        return sprintf('DROP TABLE %s', $this->tableName);
+        $statements = [];
+        foreach ($this->constraints as $constraint) {
```
Comment (Treggats) [inline src/Tempest/Database/QueryStatements/DropTableStatement.php:23]: let statements prepend the drop table, such as dropping foreign keys

File: src/Tempest/Database/QueryStatements/ForeignKeyStatement.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Database\QueryStatements;
+
+use Tempest\Database\DatabaseDialect;
+use Tempest\Database\QueryStatement;
+
+final readonly class ForeignKeyStatement implements QueryStatement
```
Comment (Treggats) [inline src/Tempest/Database/QueryStatements/ForeignKeyStatement.php:10]: not used yet, as I'm not happy with the current state

File: src/Tempest/Database/QueryStatements/IndexStatement.php [hunk 8]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Database\QueryStatements;
+
+use Tempest\Database\DatabaseDialect;
+use Tempest\Database\QueryStatement;
+
+final readonly class IndexStatement implements QueryStatement
```
Comment (Treggats) [inline src/Tempest/Database/QueryStatements/IndexStatement.php:10]: I'm unsure about simple statements like this to have a separate class

File: src/Tempest/Framework/Testing/IntegrationTestCase.php [hunk 9]
```diff
@@ -18,7 +18,7 @@
 use Tempest\Http\Method;
 use Tempest\Http\Request;
 
-abstract class IntegrationTest extends TestCase
+abstract class IntegrationTestCase extends TestCase
```
Comment (Treggats) [inline src/Tempest/Framework/Testing/IntegrationTestCase.php:21]: I think that this better indicates what kind of class this is

File: tests/Unit/Database/DatabaseQueryStatementTest.php [hunk 10]
```diff
@@ -25,127 +23,111 @@ final class DatabaseQueryStatementTest extends TestCase
...
+    public function it_can_create_an_alter_table_add_statement(DatabaseDialect $dialect, string $operation, string $validSql): void
     {
-        $this->markTestIncomplete('TODO: add alter table statement');
-        //        $statement = $driver->dialect()->createQueryStatement('Author')
-        //            ->alterTable($operation)
-        //            ->createColumn('name', 'VARCHAR(255)');
-        //
-        //        $this->assertSame($validSql, (string) $statement);
+        $query = (new AlterTableStatement('Author'))
+            ->$operation(new VarcharStatement('name'))
```
Comment (Treggats) [inline tests/Unit/Database/DatabaseQueryStatementTest.php:90]: this might be a bit magical, but it allows us to use the referenced method from the dataprovider

File: .editorconfig
Comment (aidan-casey) [inline .editorconfig:1]: I think we should probably move this to a separate PR.

**Comment by brendt** [comment]
One point of feedback: I wouldn't test on the generated text. Rather I'd execute the query into a clean DB to make sure it actually works. It's a much more robust test that doesn't fail when we make formatting changes to our queries. Here's an example: https://github.com/tempestphp/tempest-framework/blob/query-statement-defaults/tests/Integration/Database/QueryStatements/CreateTableStatementTest.php

**Comment by Treggats** [comment]
> One point of feedback: I wouldn't test on the generated text. Rather I'd execute the query into a clean DB to make sure it actually works.

Oh I like that, good idea

File: src/Tempest/Database/QueryStatements/AlterTableStatement.php [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    public function update(QueryStatement $statement): self
+    {
+        $this->statements[] = new AlterActionStatement(AlterStatement::UPDATE, $statement);
+
+        return $this;
+    }
+
+    public function delete(QueryStatement $statement): self
```
Comment (brendt) [inline src/Tempest/Database/QueryStatements/AlterTableStatement.php:32]: Shouldn't the input here simply be `string $name` instead of another statement? For `add` I understand we need it for adding columns, but I don't think it's needed for deleting them?

File: src/Tempest/Database/QueryStatements/IndexStatement.php (hunk 8 above)
Comment (brendt) [reply src/Tempest/Database/QueryStatements/IndexStatement.php:10]: I think it's ok in this case

File: src/Tempest/Framework/Testing/IntegrationTestCase.php (hunk 9 above)
Comment (brendt) [reply src/Tempest/Framework/Testing/IntegrationTestCase.php:21]: Let's keep it as a separate PR though :)

File: tests/Unit/Database/DatabaseQueryStatementTest.php (hunk 10 above)
Comment (brendt) [reply tests/Unit/Database/DatabaseQueryStatementTest.php:90]: It's ok :)

File: tests/Unit/Database/DatabaseQueryStatementTest.php [hunk 12]
```diff
@@ -25,127 +23,111 @@ final class DatabaseQueryStatementTest extends TestCase
...
+            'add',
             'ALTER TABLE Author ADD COLUMN name VARCHAR(255) NOT NULL;',
         ];
 
         yield 'mysql delete statement' => [
-            new MySqlDriver(),
-            'DELETE',
+            DatabaseDialect::MYSQL,
+            'delete',
             'ALTER TABLE Author DELETE name VARCHAR(255) NOT NULL;',
```
Comment (brendt) [inline tests/Unit/Database/DatabaseQueryStatementTest.php:119]: Yeah we don't need the whole column definition here, just the column name

**Review by brendt — CHANGES_REQUESTED** [review]

**Comment by Treggats** [comment]
> One point of feedback: I wouldn't test on the generated text. Rather I'd execute the query into a clean DB to make sure it actually works. It's a much more robust test that doesn't fail when we make formatting changes to our queries. Here's an example: [`query-statement-defaults`/tests/Integration/Database/QueryStatements/CreateTableStatementTest.php](https://github.com/tempestphp/tempest-framework/blob/query-statement-defaults/tests/Integration/Database/QueryStatements/CreateTableStatementTest.php?rgh-link-date=2024-08-17T06%3A28%3A24Z)

@brendt just made a test like this, can't use the word `table` as table name 😋

**Comment by brendt** [comment]
You can use `table` if you wrap it in backticks ;)

**Comment by Treggats** [comment]
> You can use `table` if you wrap it in backticks ;)

True, 😅 though personally I'm not very fond of backticks in programming. I remember something about `eval(..)` in PHP 😋

**Comment by brendt** [comment]
No but it should be within the compiled query, check out the CreateTableStatement where I added it as well ;)

**Comment by brendt** [comment]
Gonna merge this one in a separate branch and work on a bit more
